Sources/Apollo/GraphQLOperation.swift

```diff
@@ -1,12 +1,14 @@
-public protocol GraphQLOperation: class {
```

Comment:
Hmmm, actually, I don't think links ever need knowledge of the associated type. Here is an alternative idea: 

We keep `GraphQLOperation` as is (so without `GraphQLOperationBase`), but we define a new `GraphQLRequest` protocol and make it the basis of links. So instead of `request(operation:context:)` it would now be `perform(request:)`. 

That means `GraphQLRequest` also contains the context, basically everything needed to perform a request. Something like:

```swift
protocol GraphQLRequest {
  var context: GraphQLRequestContext { get }
  var operationType: GraphQLOperationType { get }
  var requestString: String { get }
  var variables: GraphQLMap? { get }
}
```

 I like the idea that it makes links `GraphQLRequest -> GraphQLResponse`, and it cleanly differentiates between operations (reusable across requests) and requests (containing everything that is needed fora single request, including context)

We could then have a `GraphQLOperationRequest` class that initialises itself based on a `GraphQLOperation`. So similar to your `LinkOperationDescriptor`, but with instance properties and also containing the context.

An added benefit is that it also allows you to use links without generated code, or even without the rest of Apollo iOS by implementing something like a `GraphQLDynamicRequest` class that isn't initialised from a `GraphQLOperation`.

We do need to make `GraphQLResponse` depend on `GraphQLRequest` instead of `GraphQLOperation` as well, but that might be a good idea anyway because it no longer has to be generic and that way we can more easily store responses.

What do you think?

Reply:
Sound great!